src/main/java/com/gtnewhorizon/gtnhlib/item/ItemTransfer.java

```diff
@@ -0,0 +1,201 @@
+
+    @Setter
+    protected Consumer<ItemStack> rejectedStacks;
+
+    public void source(ItemSource source) {
+        this.source = source;
+    }
+
+    public void source(Object source, ForgeDirection side) {
+        this.source = ItemUtil.getItemSource(source, side);
+    }
+
+    public void sink(ItemSink sink) {
+        this.sink = sink;
+    }
+
+    public void sink(Object sink, ForgeDirection side) {
+        this.sink = ItemUtil.getItemSink(sink, side);
+    }
+
+    public <Coord extends IBlockPos & IWorldReferent> void push(Coord pos, ForgeDirection side) {
+        TileEntity self = pos.getWorld().getTileEntity(pos.getX(), pos.getY(), pos.getZ());
+        TileEntity adjacent = pos.getWorld()
+                .getTileEntity(pos.getX() + side.offsetX, pos.getY() + side.offsetY, pos.getZ() + side.offsetZ);
+
+        push(self, side, adjacent);
+    }
+
+    public void push(Object self, ForgeDirection side, Object target) {
+        source(self, side);
+        sink(target, side.getOpposite());
+    }
+
+    public <Coord extends IBlockPos & IWorldReferent> void pull(Coord pos, ForgeDirection side) {
+        TileEntity self = pos.getWorld().getTileEntity(pos.getX(), pos.getY(), pos.getZ());
+        TileEntity adjacent = pos.getWorld()
+                .getTileEntity(pos.getX() + side.offsetX, pos.getY() + side.offsetY, pos.getZ() + side.offsetZ);
+
+        pull(self, side, adjacent);
+    }
+
+    public void pull(Object self, ForgeDirection side, Object target) {
+        source(target, side.getOpposite());
+        sink(self, side);
+    }
+
+    public void dropItems(World world, IBlockPos pos) {
+        this.itemDropping = new DroppingItemSink(world, pos);
+        if (this.rejectedStacks == null) {
+            this.rejectedStacks = itemDropping::store;
+        }
+    }
+
+    public <Coord extends IBlockPos & IWorldReferent> void dropItems(Coord pos, ForgeDirection output) {
+        dropItems(pos.getWorld(), pos.offset(output));
+    }
+
+    public void setSourceSlots(int... sourceSlots) {
+        this.sourceSlots = sourceSlots;
+    }
+
+    public void setSinkSlots(int... sinkSlots) {
+        this.sinkSlots = sinkSlots;
+    }
+
+    public int transfer() {
+        ItemSink sink = this.sink == null ? itemDropping : this.sink;
+
+        if (source == null) return 0;
+        if (sink == null) return 0;
+        if (stacksToTransfer == 0) return 0;
+        if (maxItemsPerTransfer == 0) return 0;
+
+        source.resetSource();
+        sink.resetSink();
+
+        source.setAllowedSourceSlots(sourceSlots);
+        sink.setAllowedSinkSlots(sinkSlots);
+
+        InventoryIterator iter = source.sourceIterator();
+
+        // Don't bother supporting iterator-less sources, it'll just be a performance and logic nightmare
+        if (iter == null) return 0;
+
+        int itemsTransferred = 0, stacksTransferred = 0;
+
+        outer: while (iter.hasNext() && stacksTransferred < stacksToTransfer
```

Comment:
Why not use source.pull here? Seems much easier to not use the iterator here but just pull and push. Although we should probably make sure to not always recreate the sinks iterator...